### ISSUE TITLE
Incorrect diagnostic span on syntax error in macro invoc inside trait, impl, extern blocks

### ISSUE THREAD

**Issue by abonander** [rust-lang/rust#54841]
A syntax error produced by a macro invocation in a `trait {}` or `impl {}` block will have an incorrect span for its second label:

[Playground](https://play.rust-lang.org/?gist=7a0fd6c77f665d89f5b40ec949e9b93f&version=stable&mode=debug&edition=2015)
```rust
macro_rules! m {
    () => {
        let
    }
}

trait Foo {
    m!();
}
```

Produces this error:
```
error: expected one of `async`, `const`, `extern`, `fn`, `type`, or `unsafe`, found `let`
 --> src/lib.rs:3:9
  |
1 | macro_rules! m {
  | - expected one of `async`, `const`, `extern`, `fn`, `type`, or `unsafe` here
2 |     () => {
3 |         let
  |         ^^^ unexpected token
...
8 |     m!();
  |     ----- in this macro invocation
```

The "expected one of ... here" line is pointing to line 1 but it should be elaborating on the "unexpected token" label, or just replacing it. The behavior is identical for `impl {}` blocks.

This appears to be due to some faulty logic in [`Parser::expect_one_of()`](https://github.com/rust-lang/rust/blob/master/src/libsyntax/parse/parser.rs#L681) which isn't robust in the face of macro invocations. If it can't get lines for either span from the sourcemap it should assume they're the same (in fact I'm not sure why it emits a separate "unexpected token" label).

Discovered in #54833 but not introduced by it; when that's merged `extern {}` blocks will show this behavior too (as seen in `issue-54441.stderr` on that PR).

I'll try to look into this sometime this weekend but I'm also willing to mentor on it, though like I mentioned I'm not sure why the error handling logic in `Parser::expect_one_of()` is so complex to begin with.

cc @petrochenkov

**Comment by estebank** [comment]
This is because when [looking for the prior span in the current context](https://github.com/rust-lang/rust/blob/master/src/libsyntax/parse/parser.rs#L717-L733) in order to [check wether the prior span is in a different line](https://github.com/rust-lang/rust/blob/master/src/libsyntax/parse/parser.rs#L760-L774) the code doesn't account for the case where we're in a macro context, making `self.prev_span == DUMMY_SP`, which is followed by the first char in the source code. Because of this, [we can change the following code](https://github.com/rust-lang/rust/blob/master/src/libsyntax/parse/parser.rs#L759-L779):

```
            match (cm.lookup_line(self.span.lo()), cm.lookup_line(sp.lo())) {
                (Ok(ref a), Ok(ref b)) if a.line == b.line => {
                    // When the spans are in the same line, it means that the only content between
                    // them is whitespace, point at the found token in that case:
                    //
                    // X |     () => { syntax error };
                    //   |                    ^^^^^ expected one of 8 possible tokens here
                    //
                    // instead of having:
                    //
                    // X |     () => { syntax error };
                    //   |                   -^^^^^ unexpected token
                    //   |                   |
                    //   |                   expected one of 8 possible tokens here
                    err.span_label(self.span, label_exp);
                }
                _ => {
                    err.span_label(sp, label_exp);
                    err.span_label(self.span, "unexpected token");
                }
            }
```

to 

```
            match (cm.lookup_line(self.span.lo()), cm.lookup_line(sp.lo())) {
                (Ok(ref a), Ok(ref b)) if a.line == b.line => {
                    // When the spans are in the same line, it means that the only content between
                    // them is whitespace, point at the found token in that case:
                    //
                    // X |     () => { syntax error };
                    //   |                    ^^^^^ expected one of 8 possible tokens here
                    //
                    // instead of having:
                    //
                    // X |     () => { syntax error };
                    //   |                   -^^^^^ unexpected token
                    //   |                   |
                    //   |                   expected one of 8 possible tokens here
                    err.span_label(self.span, label_exp);
                }
                _ if self.prev_span == syntax_pos::DUMMY_SP => {
                    // Account for macro context where the previous span might not be
                    // available to avoid incorrect output (#54841).
                    err.span_label(self.span, "unexpected token");
                }
                _ => {
                    err.span_label(sp, label_exp);
                    err.span_label(self.span, "unexpected token");
                }
            }
```

**Comment by abonander** [comment]
@estebank that will just print "unexpected token", no?

**Comment by abonander** [comment]
@estebank something I'm not sure on is why we print "unexpected token" at all and not just the "expected ... here" bit, or actually vice-versa. Is it for syntax errors spanning multiple tokens/lines?

**Comment by holmgr** [comment]
Hi this sounds like a chance for me to do my first contribution, do you mind if I take it?

**Comment by abonander** [comment]
@holmgr please do! If you need help just let me know.

**Comment by holmgr** [comment]
Sorry for the time taken but I have managed to implement the change suggested by @estebank [here](https://github.com/holmgr/rust/commit/95aed9862c06fa2dcb6eca592e91680ca170466d) and confirm that the output matches:

```
error: expected one of `async`, `const`, `extern`, `fn`, `type`, or `unsafe`, found `let`
 --> test.rs:3:9
  |
3 |         let
  |         ^^^ unexpected token
...
8 |     m!();
  |     ----- in this macro invocation

error: aborting due to previous error
```

I did modify the only test affected.
In that case I guess I should just open a PR referencing this issue?

**Comment by estebank** [comment]
Yes go ahead and file a pr with your change. Thanks for fixing this!

**Comment by holmgr** [comment]
The PR is now merged, so I guess we could go ahead and close this issue now
